packages/gossip/neighbor.go

```diff
@@ -33,149 +35,130 @@ const (
 }
 
 // NewNeighbor creates a new neighbor from the provided peer and connection.
-func NewNeighbor(p *peer.Peer, group NeighborsGroup, conn net.Conn, log *logger.Logger) *Neighbor {
-	if !IsSupported(p) {
-		panic("peer does not support gossip")
-	}
-
-	// always include ID and address with every log message
+func NewNeighbor(p *peer.Peer, group NeighborsGroup, stream network.Stream, log *logger.Logger) *Neighbor {
 	log = log.With(
 		"id", p.ID(),
-		"network", conn.LocalAddr().Network(),
-		"addr", conn.RemoteAddr().String(),
+		"localAddr", stream.Conn().LocalMultiaddr(),
+		"remoteAddr", stream.Conn().RemoteMultiaddr(),
 	)
-
 	return &Neighbor{
-		Peer:                  p,
-		Group:                 group,
-		BufferedConnection:    buffconn.NewBufferedConnection(conn, maxPacketSize),
-		log:                   log,
-		queue:                 make(chan []byte, neighborQueueSize),
-		closing:               make(chan struct{}),
-		connectionEstablished: time.Now(),
+		Peer:  p,
+		Group: group,
+
+		log: log,
+
+		stream: stream,
+		reader: libp2putil.NewDelimitedReader(stream),
+		writer: libp2putil.NewDelimitedWriter(stream),
+
+		disconnected:   events.NewEvent(disconnected),
+		packetReceived: events.NewEvent(packetReceived),
+		packetsRead:    atomic.NewUint64(0),
+		packetsWritten: atomic.NewUint64(0),
 	}
 }
 
-// ConnectionEstablished returns the connection established.
-func (n *Neighbor) ConnectionEstablished() time.Time {
-	return n.connectionEstablished
+// PacketsRead returns number of packets this neighbor has received.
+func (n *Neighbor) PacketsRead() uint64 {
+	return n.packetsRead.Load()
 }
 
-// Listen starts the communication to the neighbor.
-func (n *Neighbor) Listen() {
-	n.wg.Add(2)
-	go n.readLoop()
-	go n.writeLoop()
-
-	n.log.Info("Connection established")
+// PacketsWritten returns number of packets this neighbor has sent.
+func (n *Neighbor) PacketsWritten() uint64 {
+	return n.packetsWritten.Load()
 }
 
-// Close closes the connection to the neighbor and stops all communication.
-func (n *Neighbor) Close() error {
-	err := n.disconnect()
-	// wait for everything to finish
-	n.wg.Wait()
-
-	return err
+func disconnected(handler interface{}, _ ...interface{}) {
+	handler.(func())()
 }
 
-// IsOutbound returns true if the neighbor is an outbound neighbor.
-func (n *Neighbor) IsOutbound() bool {
-	return GetAddress(n.Peer) == n.RemoteAddr().String()
+func packetReceived(handler interface{}, params ...interface{}) {
+	handler.(func(*pb.Packet))(params[0].(*pb.Packet))
 }
 
-func (n *Neighbor) disconnect() (err error) {
-	n.disconnectOnce.Do(func() {
-		close(n.closing)
-		err = n.BufferedConnection.Close()
-		n.log.Info("Connection closed")
-	})
-	return
+// ConnectionEstablished returns the connection established.
+func (n *Neighbor) ConnectionEstablished() time.Time {
+	return n.stream.Stat().Opened
 }
 
-func (n *Neighbor) writeLoop() {
+func (n *Neighbor) readLoop() {
+	n.wg.Add(1)
 	defer n.wg.Done()
```

Comment:
Shouldn't the defer go inside the go routine?

Reply:
Good catch!